### PR TITLE
Publisher#range operator

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The canonocal way to do an asynchornous for lop with integer indexes in reactive streams is to use a range() operator. We currently don't have a range operator.

Modifications:
- Add Publisher#range(int begin, int end)
- Add Publisher#range(int begin, int end, int stride)

Result:
For loops with integer indexes can be done directly with Publisher#range(..)

### CODE REVIEW

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    @Override
+    void doSubscribe(final Subscriber<? super Integer> subscriber) {
+        subscriber.onSubscribe(new RangeIntSubscription(subscriber));
+    }
+
+    private final class RangeIntSubscription implements PublisherSource.Subscription {
+        private final PublisherSource.Subscriber<? super Integer> subscriber;
+        private long pendingN;
+        private int index = RangeIntPublisher.this.begin;
```
Comment (normanmaurer) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java]: nit: consider passing begin into the constructor so you can make the class static.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            pendingN = addWithOverflowProtection(pendingN, n);
+            for (; pendingN > 0 && index < end; --pendingN, index += min(stride, (long) end - index)) {
+                try {
+                    subscriber.onNext(index);
+                } catch (Throwable cause) {
+                    sendOnError(cause);
+                    return;
+                }
+            }
+            if (index >= end) {
```
Comment (normanmaurer) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java]: nit:  `==` should be enough ?

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    RangeIntPublisher(int begin, int end) {
+        this(begin, end, 1);
+    }
+
+    RangeIntPublisher(int begin, int end, int stride) {
+        if (begin > end) {
+            throw new IllegalArgumentException("begin(" + begin + ") > end(" + end + ")");
+        }
+        if (stride <= 0) {
+            throw new IllegalArgumentException("stride: " + stride);
```
[the remark this answers is not in the document]
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java]: done

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java (hunk 1 above)
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java]: the inner class also needs to access `end` and `stride` from the outer class. I will leave as is for now.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java (hunk 2 above)
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RangeIntPublisher.java]: sure, let me update.

File: servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/publisher/FromIntRangePublisherTest.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    public void reentrantCancelStopsDelivery() {
+        reentrantCancelStopsDelivery(true);
+    }
+
+    @Test
+    public void reentrantCancelStopsDeliveryStride() {
+        reentrantCancelStopsDelivery(false);
+    }
+
+    private void reentrantCancelStopsDelivery(boolean stride) {
```
Comment (NiteshKant) [inline servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/publisher/FromIntRangePublisherTest.java]: An additional test case may be useful to verify `request-n` post `cancel` are ignored.

**Pull request merged**